src/services/spot/Spot.cpp

```diff
@@ -0,0 +1,387 @@
+
+#include "rapidjson/rapidjson.h"
+#include "rapidjson/document.h"
+#include "rapidjson/ostreamwrapper.h"
+#include "rapidjson/writer.h"
+
+#include <iostream>
+#include <sstream>
+#include <iterator>
+#include <fstream>
+#include <time.h>
+using namespace cali;
+
+namespace
+{
+
+    class Spot {
+        static std::unique_ptr<Spot> s_instance;
+        static const ConfigSet::Entry  s_configdata[];
+        static int divisor;
+        template<typename T>
+        using List = std::vector<T>;
+        using AggregationHandler = cali::Aggregator*;
+        using SelectionHandler = cali::RecordSelector*;
+        using TimeType = unsigned int;
+        using SingleJsonEntryType = List<std::pair<std::string,TimeType>>;
+        using JsonListType = List<SingleJsonEntryType>;
+        using AggregationDescriptor = std::pair<std::string, std::string>;
+        using AggregationDescriptorList = List<AggregationDescriptor>;
+        static List<std::pair<AggregationHandler,SelectionHandler>> m_queries;
+        static std::vector<std::string> y_axes;
+        static AggregationDescriptorList m_annotations_and_places;
+        static JsonListType m_jsons;
+        static std::string code_version;
+        static std::string recorded_time;
+        static std::vector<std::string> title;
+        static bool include_local;
+        static bool show_exclusive;
+        void process_snapshot(Caliper* c, const SnapshotRecord* snapshot) {
+          for(auto& m_query : m_queries){
+            auto entries = snapshot->to_entrylist();
+            if(m_query.second->pass(*c,entries)){
+              m_query.first->add(*c, entries);
+            }
+          }
+        }
+        std::string extract_parent_name(CaliperMetadataAccessInterface& db,const Node* node,std::string metric = "time.inclusive.duration",bool initcall=false){
+          if((!node) ||(!node->attribute()) ){
+            return "";
+          }
+          if(db.get_attribute(node->attribute()).name()!=metric){
+            return extract_parent_name(db,node->parent(),metric);
+          }
+          std::string my_name = node->data().to_string();
+          if(my_name.size()>0){
+            std::string parent_name = extract_parent_name(db,node->parent(),metric);
+            return parent_name+my_name+(initcall ? "" : "/");
+          }
+          return my_name;
+        } 
+        void flush(Caliper* c, const SnapshotRecord*) {
+          for(int i =0 ;i<m_queries.size();i++) {
+            auto& m_query = m_queries[i];
+            auto& m_json = m_jsons[i];
+            std::string grouping = m_annotations_and_places[i].first;
+            std::string end_grouping = "event.end#"+grouping;
+            std::vector<std::string> metrics_of_interest { "time.inclusive.duration", 
+              grouping, end_grouping};
+            m_query.first->flush(*c,[&](CaliperMetadataAccessInterface& db,const EntryList& list) {
+                std::string name;
```

Comment:
This whole thing seems like a bit odd. It seems you just want to extract the hierarchy for the "grouping" attribute plus the time?